x/oracle/handler.go

```diff
@@ -22,41 +27,98 @@ func NewHandler(k Keeper) sdk.Handler {
 	}
 }
 
-// handleMsgPriceFeed handles a MsgPriceFeed
-func handleMsgPriceFeed(ctx sdk.Context, keeper Keeper, pfm MsgPriceFeed) sdk.Result {
+// handleMsgPricePrevote handles a MsgPricePrevote
+func handleMsgPricePrevote(ctx sdk.Context, keeper Keeper, ppm MsgPricePrevote) sdk.Result {
+	valset := keeper.valset
+
+	if !ppm.Feeder.Equals(ppm.Validator) {
+		delegate := keeper.GetFeedDelegate(ctx, ppm.Validator)
+		if !delegate.Equals(ppm.Feeder) {
+			return ErrNoVotingPermission(DefaultCodespace, ppm.Feeder, ppm.Validator).Result()
+		}
+	}
+
+	// Check that the given validator exists
+	val := valset.Validator(ctx, ppm.Validator)
+	if val == nil {
+		return staking.ErrNoValidatorFound(DefaultCodespace).Result()
+	}
+
+	prevote := NewPricePrevote(ppm.Hash, ppm.Denom, ppm.Validator, ctx.BlockHeight())
+	keeper.addPrevote(ctx, prevote)
+
+	return sdk.Result{
+		Tags: sdk.NewTags(
+			tags.Denom, ppm.Denom,
+			tags.Voter, ppm.Validator.String(),
+			tags.FeedDelegate, ppm.Feeder.String(),
+		),
+	}
+}
+
+// handleMsgPriceVote handles a MsgPriceVote
+func handleMsgPriceVote(ctx sdk.Context, keeper Keeper, pvm MsgPriceVote) sdk.Result {
 	valset := keeper.valset
 
-	if !pfm.Feeder.Equals(pfm.Validator) {
-		delegate := keeper.GetFeedDelegate(ctx, pfm.Validator)
-		if !delegate.Equals(pfm.Feeder) {
-			return ErrNoVotingPermission(DefaultCodespace, pfm.Feeder, pfm.Validator).Result()
+	if !pvm.Feeder.Equals(pvm.Validator) {
```

Comment:
Can pvm.Feeder ever equal pvm.Validator? the former is of the type sdk.AccAddress and the second is of the type sdk.ValAddress

Reply:
`Equals` function will compare Bytes of each variable. So I can equal for same byte addresses.